Sources/AsyncHTTPClient/ConnectionPool/HTTP2/HTTP2Connection.swift

```diff
@@ -0,0 +1,146 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Logging
+import NIO
+import NIOHTTP2
+
+protocol HTTP2ConnectionDelegate {
+    func http2ConnectionStreamClosed(_: HTTP2Connection, availableStreams: Int)
+    func http2ConnectionClosed(_: HTTP2Connection)
+}
+
+class HTTP2Connection {
+    let channel: Channel
+    let multiplexer: HTTP2StreamMultiplexer
+    let logger: Logger
+
+    /// the connection pool that created the connection
+    let delegate: HTTP2ConnectionDelegate
+
+    enum State {
+        case starting(EventLoopPromise<Void>)
+        case active(HTTP2Settings)
+        case closed
+    }
+
+    var readyToAcceptConnectionsFuture: EventLoopFuture<Void>
+
+    var settings: HTTP2Settings? {
+        self.channel.eventLoop.assertInEventLoop()
+        switch self.state {
+        case .starting:
+            return nil
+        case .active(let settings):
+            return settings
+        case .closed:
+            return nil
+        }
+    }
+
+    private var state: State
+    let id: HTTPConnectionPool.Connection.ID
+
+    init(channel: Channel,
+         connectionID: HTTPConnectionPool.Connection.ID,
+         delegate: HTTP2ConnectionDelegate,
+         logger: Logger) throws {
+        precondition(channel.isActive)
+        channel.eventLoop.preconditionInEventLoop()
+
+        let readyToAcceptConnectionsPromise = channel.eventLoop.makePromise(of: Void.self)
+
+        self.channel = channel
+        self.id = connectionID
+        self.logger = logger
+        self.multiplexer = HTTP2StreamMultiplexer(
+            mode: .client,
+            channel: channel,
+            targetWindowSize: 65535,
+            outboundBufferSizeHighWatermark: 8196,
+            outboundBufferSizeLowWatermark: 4092,
+            inboundStreamInitializer: { (channel) -> EventLoopFuture<Void> in
+                struct HTTP2PushNotsupportedError: Error {}
+                return channel.eventLoop.makeFailedFuture(HTTP2PushNotsupportedError())
+            }
+        )
+        self.delegate = delegate
+        self.state = .starting(readyToAcceptConnectionsPromise)
+        self.readyToAcceptConnectionsFuture = readyToAcceptConnectionsPromise.futureResult
+
+        // 1. Modify channel pipeline and add http2 handlers
+        let sync = channel.pipeline.syncOperations
+
+        let http2Handler = NIOHTTP2Handler(mode: .client, initialSettings: nioDefaultSettings)
+        let idleHandler = HTTP2IdleHandler(connection: self, logger: self.logger)
+
+        try sync.addHandler(http2Handler, position: .last)
+        try sync.addHandler(idleHandler, position: .last)
+        try sync.addHandler(self.multiplexer, position: .last)
```

Comment:
Why are we building the HTTP/2 handlers ourselves?

Reply:
Because we need to add the IdleHandler in the middle. This trick was imported from grpc based on @glbrntt's suggestion.

Reply:
Can we add a comment to that effect please?